docs(kv-ir): Add background necessary to understand the kv-ir stream format. (#833)
y-scope/clp: docs/src/dev-guide/design-kv-ir-streams/background.md

| StructuredArray   | An array                                                                                               | Internal  |

+++
**Table 3**: clp-s value types.
:::

(figure-2)=
::::{card}
:::{mermaid}
%%{
  init: {
    "theme": "base",
    "themeVariables": {
      "primaryColor": "#0066cc",
      "primaryTextColor": "#fff",
      "primaryBorderColor": "transparent",
      "lineColor": "#007fff",
      "secondaryColor": "#007fff",
      "tertiaryColor": "#fff"
    }
  }
}%%
flowchart LR
  rootObj("root: <span style='color: #97ff00'>Object</span>")
  timestampInt("timestamp: <span style='color: #97ff00'>Integer</span>")
  levelVarStr("level: <span style='color: #97ff00'>VarString</span>")
  messageClpStr("message: <span style='color: #97ff00'>ClpString</span>")
  timersObj("timers: <span style='color: #97ff00'>Object</span>")
  timersStage1Float("stage_1: <span style='color: #97ff00'>Float</span>")
  timersStage2Null("stage_2: <span style='color: #97ff00'>NullValue</span>")

  rootObj --> timestampInt
  rootObj --> levelVarStr
  rootObj --> messageClpStr
  rootObj --> timersObj
  timersObj --> timersStage1Float
  timersObj --> timersStage2Null
:::

> +++

Author: How can I use Mermaid's theme variables or annotations?

# Background

To understand the KV-IR stream format, we first need to review the following:

* [How clp-s compresses log events](#clp-s-compression), since the process for KV-IR streams is
  similar but makes a different trade-off between resource usage, compression ratio, and search
  performance.
* [How clp-s parses and encodes unstructured text values](#parsing--encoding-unstructured-text).

We discuss each below.

## clp-s compression

At a high-level, [clp-s](../../user-guide/core-clp-s.md) compresses log events into what we call
archives. Depending on the configured size-threshold for each archive, a set of log events may be

> compressed into one or more archives. In addition, the format of an archive is designed so that each
> archive is independent of other archives, meaning that different archives can be searched
> concurrently.

Reviewer: ```suggestion
compressed into one or more archives. In addition, the format of an archive is designed to be
self-contained and independent, allowing archives to be searched concurrently.
```
Trying to make it more concise.

and [Table 2](#table-2). clp-s' value types, including those used in the schema, are listed in
[Table 3](#table-3).

clp-s assigns a type to a value based on the value's "abstract" type (i.e., whether it's an integer,
float, boolean, string, object, null, or array) and how the value should be encoded. For some
abstract types, clp-s only has one way of encoding it, so it assigns the corresponding clp-s
type--e.g., the integer corresponding to the `timestamp` key. For other abstract types, clp-s can
encode the value in multiple ways, so it assigns the clp-s type where the encoded value will result
in a good trade-off between compactness and efficient searches---e.g., the strings corresponding to
the `level` and `message` keys use different clp-s types.

Once clp-s assigns a type to a value, it can add it to the log event's schema tree. Except for the
root, each node in a schema tree represents a unique key and value-type pair from the schema. For
instance, the tree for the schema in [Table 1](#table-1) is shown in [Figure 2](#figure-2). Since
the tree represents the structure of a structured log event, the internal (non-leaf) nodes will

> always correspond to `Object`s or `StructuredArray`s, while the leaf nodes will correspond to values

Reviewer: ```suggestion
always correspond to `Object` or `StructuredArray`, while the leaf nodes will correspond to values
```
I'm not sure if plural form here is necessary... There're several more in the below sections.

Author: For some instances, I can rephrase it to be singular, but for others, plurals make more sense. I guess your concern is with how it renders as `Object`s? If so, I can rewrite it to `Object` values (for instance).

Reviewer: Right. "`Object` values" sgtm.

      "primaryTextColor": "#fff",
      "primaryBorderColor": "transparent",
      "lineColor": "#007fff",
      "secondaryColor": "#007fff",
      "tertiaryColor": "#fff"
    }
  }
}%%
flowchart LR
  rootObj("root: <span style='color: #97ff00'>Object</span>")
  timestampInt("timestamp: <span style='color: #97ff00'>Integer</span>")
  levelVarStr("level: <span style='color: #97ff00'>VarString</span>")
  messageClpStr("message: <span style='color: #97ff00'>ClpString</span>")
  timersObj("timers: <span style='color: #97ff00'>Object</span>")
  timersStage1Float("stage_1: <span style='color: #97ff00'>Float</span>")
  timersStage2Null("stage_2: <span style='color: #97ff00'>NullValue</span>")

  rootObj --> timestampInt
  rootObj --> levelVarStr
  rootObj --> messageClpStr
  rootObj --> timersObj
  timersObj --> timersStage1Float
  timersObj --> timersStage2Null
:::
+++
**Figure 2**: The schema tree for log event &#35;1 in [Figure 1](#figure-1). Each node's label is of
the form `<key>: <type>`, and each arrow is from a parent to a child node.
::::

### Encoding log event schemas

To compactly encode the schemas of the events in an archive, clp-s maintains an archive-level schema

> tree that merges all events' schema trees; each event's schema can then be uniquely encoded as
> an array of identifiers for the relevant leaf nodes of the archive's tree. (The leaf node IDs are
> sufficient since we can determine the predecessor nodes by traversing to the root from the leaves.)

Reviewer: ```suggestion
tree that merges all events' schema trees. Each event's schema can then be uniquely encoded as
an array of identifiers for the relevant leaf nodes of the archive's tree, since the leaf node IDs
are sufficient as we can determine the predecessor nodes by traversing to the root from the leaves.
```
Try to break the first sentence from the second one, and remove `()` to make the sentence more complete.

For instance, [Figure 3](#figure-3) shows the schema tree after adding the example logs
([Figure 1](#figure-1)) to the tree. Referencing the leaf node IDs, the schema for event &#35;1 can
be encoded as `[1, 2, 3, 5, 7]`. To merge an event's schema tree with the archive's tree, clp-s
iterates over each pair of nodes---one from each tree:

* If the nodes have the same key and value-type, and all of their predecessor nodes have matching
  key and value-type pairs, clp-s merges the nodes in the resulting tree.
* Otherwise, both nodes are added to the resulting tree, and each is assigned a unique integer ID.

<!-- markdownlint-disable MD013 -->
(figure-3)=
::::{card}
:::{mermaid}
%%{
  init: {
    "theme": "base",
    "themeVariables": {
      "primaryColor": "#0066cc",
      "primaryTextColor": "#fff",
      "primaryBorderColor": "transparent",
      "lineColor": "#007fff",
      "secondaryColor": "#007fff",
      "tertiaryColor": "#fff"
    }
  }
}%%
flowchart LR

>   rootObj("<span style='color: #ffbe00'>0</span> root: <span style='color: #97ff00'>Object</span>")
>   timestampInt("<span style='color: #ffbe00'>1</span> timestamp: <span style='color: #97ff00'>Integer</span>")
>   levelVarStr("<span style='color: #ffbe00'>2</span> level: <span style='color: #97ff00'>VarString</span>")
>   messageClpStr("<span style='color: #ffbe00'>3</span> message: <span style='color: #97ff00'>ClpString</span>")
>   timersObj("<span style='color: #ffbe00'>4</span> timers: <span style='color: #97ff00'>Object</span>")
>   timersStage1Float("<span style='color: #ffbe00'>5</span> stage_1: <span style='color: #97ff00'>Float</span>")
>   timersStage2Null("<span style='color: #ffbe00'>6</span> stage_2: <span style='color: #97ff00'>NullValue</span>")
>   timersStage2Float("<span style='color: #ffbe00'>7</span> stage_2: <span style='color: #97ff00'>Float</span>")
> 
>   rootObj --> timestampInt
>   rootObj --> levelVarStr
>   rootObj --> messageClpStr
>   rootObj --> timersObj
>   timersObj --> timersStage1Float
>   timersObj --> timersStage2Null
>   timersObj --> timersStage2Float

Reviewer: ```suggestion
  rootObj("<span style='color: #ffbe00'>0</span> root: <span style='color: #97ff00'>Object</span>")
  timestampInt("<span style='color: #ffbe00'>1</span> timestamp: <span style='color: #97ff00'>Integer</span>")
  levelVarStr("<span style='color: #ffbe00'>2</span> level: <span style='color: #97ff00'>VarString</span>")
  timersObj("<span style='color: #ffbe00'>4</span> timers: <span style='color: #97ff00'>Object</span>")
  messageClpStr("<span style='color: #ffbe00'>3</span> message: <span style='color: #97ff00'>ClpString</span>")
  timersStage1Float("<span style='color: #ffbe00'>5</span> stage_1: <span style='color: #97ff00'>Float</span>")
  timersStage2Null("<span style='color: #ffbe00'>6</span> stage_2: <span style='color: #97ff00'>NullValue</span>")
  timersStage2Float("<span style='color: #ffbe00'>7</span> stage_2: <span style='color: #97ff00'>Float</span>")

  rootObj --> timestampInt
  rootObj --> levelVarStr
  rootObj --> timersObj
  rootObj --> messageClpStr
  timersObj --> timersStage1Float
  timersObj --> timersStage2Null
  timersObj --> timersStage2Float
```
Can we switch `timersObj` and `messageClpStr`'s order so that the diagram doesn't overlap?
![image](https://github.com/user-attachments/assets/c0568f38-d816-41ef-8dea-f136a0034093)

:::
+++
**Figure 3**: The archive's schema tree after adding the log events from [Figure 1](#figure-1). Each
node's label is of the form `<ID> <key>: <type>`.
::::
<!-- markdownlint-enable MD013 -->

### Encoding & storing event values

> For each log event, clp-s uses a variety of encoding methods to compactly encode its values before
> storing them in a table corresponding to its schema. We refer to this table as an encoded record
> table (ERT). As we'll see below, clp-s uses a different encoding method for each of its value types,
> with the goal of each encoding method being to deduplicate any repetitive information (e.g.,
> deduplicating repeated `VarString`s with a dictionary) and then represent the value with a 64-bit
> integer. For each encoded value, clp-s stores it in the value's corresponding ERT column. ERTs are
> efficient to search since all the values are integers; and by grouping events with the same schema
> into an ERT, clp-s essentially deduplicates the event's schema.

Reviewer: ```suggestion
For each log event, clp-s uses a variety of encoding methods to compactly encode its values before
storing them in a table corresponding to its schema. We refer to this table as an encoded record
table (ERT).

As we'll see below, clp-s uses different encoding methods for each of its value types, with the
primary goal of eliminating redundancy (e.g., deduplicating repeated `VarString` values via
dictionary encoding) and then converting the value into a 64-bit integer. These encoded values are
stored in their respective columns within the ERT.

ERTs are efficient to search since all the ERT columns store integers. Additionally, clp-s avoids
redundant storage of schema information by grouping events with the same schema into an ERT.
```
- Tried to break this paragraph into three short ones (personally this makes it easier to read):
  - Intro to ERT
  - How ERT stores values
  - The efficiency of this format
- Tried to shorten the sentences a little bit by rewording or eliminating `;`.

Author: I took a different approach by breaking the subsection into two---one about encoding values and one about storing them.

> [Table 4](#table-4) lists how clp-s encodes each leaf-node value type. Most value types are encoded
> conventionally, but a few require further explanation. For the values encoded as dictionary IDs,
> clp-s simply stores the value in a dictionary and maps it to a unique integer ID. For `ClpString`s,
> clp-s encodes each component separately. Finally, `NullValue`s don't need to be encoded since they
> don't need to be stored explicitly---a `NullValue` leaf node already indicates that the
> corresponding column of the ERT is null.

Reviewer: ```suggestion
[Table 4](#table-4) lists how clp-s encodes each leaf-node value type. Most value types are encoded
conventionally with the following exceptions:
- For the values encoded as dictionary IDs, clp-s simply stores the value in a dictionary and maps
  it to a unique integer ID.
- For `ClpString`s, clp-s encodes each component separately.
- For `NullValue`s, clp-s doesn't need to encode anything since they don't need to be stored
  explicitly---a `NullValue` leaf node already indicates that the corresponding column of the ERT is
  null.
```
Personally I think bullet points make this paragraph more readable

(table-4)=
:::{card}

| clp-s value-type            | Encoding                                                |
|-----------------------------|---------------------------------------------------------|
| Integer                     | 8 B integer                                             |
| Float                       | 8 B IEEE-754 double-precision float                     |
| Boolean                     | 1 B integer                                             |
| VarString                   | 8 B dictionary ID                                       |
| DateString                  | 8 B epoch timestamp and 8 B format string dictionary ID |
| ClpString                   | *See below*                                             |
| --> Format string           | 8 B dictionary ID                                       |
| --> Encoded variable values | Collection of 8 B integers                              |
| --> String variable values  | Collection of 8 B dictionary IDs                        |
| UnstructuredArray           | Same as ClpString                                       |

> | NullValue                   | N/A                                                     |

Reviewer: Shall we use `8-byte` instead of `8 B`?

For other arrays, the `UnstructuredArray` type is more appropriate---since it's encoded as a JSON
string, its elements won't be added to the tree. Nonetheless, values within these arrays can still
be searched.

### Writing archives to disk

To write an archive's data structures to disk, clp-s serializes them and writes them to one or more
general-purpose compression streams. Applying general-purpose compression allows us to mitigate some
of the inefficient encodings (e.g., encoding `Boolean`s as integers) used to maintain efficient
search performance. For some data structures, like dictionaries, clp-s writes them to disk as they
are built; yet for other data structures, like the ERTs, clp-s buffers them in memory until the
archive is complete.

## Parsing & encoding unstructured text

clp-s uses [clp](../../user-guide/core-unstructured/clp.md)'s algorithm to parse and encode
unstructured text. Unstructured text is a string that contains zero or more variable values
interspersed with non-variable (static) text. For example, in [Figure 1](#figure-1), log event
&#35;1's `message` value is unstructured text containing the variable values `task_1` and `2`. At a

> high-level, the clp algorithm uses a set of user-defined regular expressions to match each variable

Reviewer: ```suggestion
high-level, clp's algorithm uses a set of user-defined regular expressions to match each variable
```
This is to match the first sentence.

value in the unstructured text, decomposing the text into:

* a format string---i.e., the unstructured text with variable values replaced with placeholders;

> * string variable values; and

Reviewer: Do we need the `and`?

Author: As far as I know, grammatically, that's the way to write it, since all bullets are necessary together rather than being independent. That said, I think it's still understandable without the and, so I'll remove it for conciseness.

* encoded variable values---i.e., variable values which have been encoded as 64-bit integers.

Collectively, we refer to these three components as an *encoded text AST*. For instance, log event
&#35;1's `message` value would be decomposed into the following encoded text AST:

> * Format string: `\x12 completed successfully. \x11 task(s) remain.`
> * `\x12` and `\x11` are variable placeholders representing string and integer variables,
>   respectively.
> * String variable values: `["task_1"]`
> * Encoded variable values: `[1]`

Reviewer: ```suggestion
* Format string: `\x12 completed successfully. \x11 task(s) remain.`
  * `\x12` and `\x11` are variable placeholders representing string and integer variables,
    respectively.
* String variable values: `["task_1"]`
* Encoded variable values: `[1]`
```
Feel like the second one is the sub-point of the first one?

  timestampInt("<span style='color: #ffbe00'>1</span> timestamp: <span style='color: #97ff00'>Integer</span>")
  levelVarStr("<span style='color: #ffbe00'>2</span> level: <span style='color: #97ff00'>VarString</span>")
  timersObj("<span style='color: #ffbe00'>4</span> timers: <span style='color: #97ff00'>Object</span>")
  timersStage1Float("<span style='color: #ffbe00'>5</span> stage_1: <span style='color: #97ff00'>Float</span>")
  timersStage2Null("<span style='color: #ffbe00'>6</span> stage_2: <span style='color: #97ff00'>NullValue</span>")
  timersStage2Float("<span style='color: #ffbe00'>7</span> stage_2: <span style='color: #97ff00'>Float</span>")
  messageClpStr("<span style='color: #ffbe00'>3</span> message: <span style='color: #97ff00'>ClpString</span>")

  rootObj --> timestampInt
  rootObj --> levelVarStr
  rootObj --> timersObj
  timersObj --> timersStage1Float
  timersObj --> timersStage2Null
  timersObj --> timersStage2Float
  rootObj --> messageClpStr
:::
+++
**Figure 3**: The archive's schema tree after adding the log events from [Figure 1](#figure-1). Each
node's label is of the form `<ID> <key>: <type>`.
::::

> <!-- markdownlint-enable MD013 -->

Reviewer 2: I don't think its worth explaining in detail, but do you think its worth drawing the full archive level schema tree instead of just the default object namespace portion?

i.e. have a separate untyped `root`, and maybe event include the `log_event_idx` in the `Metadata` subtree?

I guess tree would also have
```
root -> rootObj
root -> metadata
metadata -> log_event_idx
```
